File: apps/scryglass/src/lib/playerMapStats.ts
/** Reader for features/player_map_stats.json, the profile Stats section source.
 *
 * The artifact is display-only and optional, so every reader here fails closed
 * to `null` rather than throwing: a release published without the asset must
 * still render both profiles, minus the Stats section.
 *
 * `players` and `teams` are arrays carrying a `name` field rather than
 * name-keyed objects. The publication guard
 * `public.scryglass_json_has_draft_fields` inspects JSON object *keys*, so a
 * name-keyed map would submit player and team names to the banned-key list.
 * Keep the array shape.
 */

import { readChatJson } from "@/lib/chatApi";
import type { PackManifest } from "@/lib/pack";
import { teamQueryAliases } from "@/lib/pack";
import { readPackJson } from "@/lib/serverPack";

export const PLAYER_MAP_STATS_PATH = "features/player_map_stats.json";
export const PLAYER_MAP_STATS_SCHEMA_VERSION = "scryglass:player-map-stats:v1";

/** One published map. Every metric is nullable: absent is never zero. */
export type MapStatsGame = {
  game_id: string;
  date: string;
  league: string | null;
  opponent: string | null;
  champion?: string | null;
  position?: string | null;
  win: boolean;
  cs_per_min?: number | null;
  gold_per_min: number | null;
  gold_share_pct?: number | null;
  damage_per_min: number | null;
  damage_share_pct?: number | null;
  kda?: number | null;
  kills?: number | null;
  deaths?: number | null;
  game_length_min: number | null;
};

/** Aggregate header over exactly the `games` rows beneath it, plus those rows. */
export type MapStatsEntry = {
  name: string;
  maps: number;
  wins: number;
  losses: number;
  win_rate: number | null;
  cs_per_min?: number | null;
  gold_per_min: number | null;
  gold_share_pct?: number | null;
  damage_per_min: number | null;
  damage_share_pct?: number | null;
  kda?: number | null;
  kills?: number | null;
  deaths?: number | null;
  game_length_min?: number | null;
  games: MapStatsGame[];
};

export type PlayerMapStats = {
  schema_version: string;
  window_days: number;
  map_limit: number;
  players: MapStatsEntry[];
  teams: MapStatsEntry[];
};

export type MapStatsKind = "players" | "teams";

export type MapStatsLookup = {
  entry: MapStatsEntry;
  window_days: number;
  map_limit: number;
};

function isNullableNumber(value: unknown): boolean {
  return value == null || Number.isFinite(value);
}

function isGame(value: unknown): value is MapStatsGame {
  if (!value || typeof value !== "object") return false;
  const game = value as Record<string, unknown>;
  return typeof game.game_id === "string"
    && game.game_id.length > 0
    && typeof game.date === "string"
    && game.date.length > 0
    && typeof game.win === "boolean"
    && (game.league == null || typeof game.league === "string")
    && (game.opponent == null || typeof game.opponent === "string")
    && isNullableNumber(game.cs_per_min)
    && isNullableNumber(game.gold_per_min)
    && isNullableNumber(game.gold_share_pct)
    && isNullableNumber(game.damage_per_min)
    && isNullableNumber(game.damage_share_pct)
    && isNullableNumber(game.kda)
    && isNullableNumber(game.game_length_min);
}

function isEntry(value: unknown): value is MapStatsEntry {
  if (!value || typeof value !== "object") return false;
  const record = value as Record<string, unknown>;
  if (typeof record.name !== "string" || record.name.length === 0) return false;
  if (!Number.isFinite(record.maps) || !Number.isFinite(record.wins) || !Number.isFinite(record.losses)) return false;
  if (!isNullableNumber(record.win_rate)) return false;
  if (!Array.isArray(record.games)) return false;
  // Fail closed on the WHOLE entry when any row is malformed. The aggregate
  // header is documented as the mean over exactly the rows beneath it, so
  // silently dropping a bad row would publish a header disagreeing with its
  // own table; and a null row previously crashed the profile at sort time.
  return record.games.every(isGame);
}

/** Accept only a payload that declares the exact schema this reader knows. */
export function parsePlayerMapStats(value: unknown): PlayerMapStats | null {
  if (!value || typeof value !== "object") return null;
  const record = value as Record<string, unknown>;
  if (record.schema_version !== PLAYER_MAP_STATS_SCHEMA_VERSION) return null;
  if (!Array.isArray(record.players) || !Array.isArray(record.teams)) return null;
  const windowDays = Number(record.window_days);
  const mapLimit = Number(record.map_limit);
  if (!Number.isFinite(windowDays) || !Number.isFinite(mapLimit)) return null;
  return {
    schema_version: PLAYER_MAP_STATS_SCHEMA_VERSION,
    window_days: windowDays,
    map_limit: mapLimit,
    players: record.players.filter(isEntry),
    teams: record.teams.filter(isEntry),
  };
}

/** Exact name first, so source casing never decides a collision. */
export function findMapStatsEntry(
  entries: readonly MapStatsEntry[],
  name: string,
  kind: MapStatsKind = "players",
): MapStatsEntry | null {
  const wanted = name.trim();
  if (!wanted) return null;
  // Teams are commonly queried by alias (DK, KC, MKOI); resolve through the
  // existing alias contract so those do not return a false not-found.
  const candidates = kind === "teams" ? teamQueryAliases(wanted) : [wanted];
  for (const candidate of candidates) {
    const exact = entries.find((entry) => entry.name === candidate);
    if (exact) return exact;
  }
  for (const candidate of candidates) {
    const lower = candidate.toLowerCase();
    const match = entries.find((entry) => entry.name.toLowerCase() === lower);
    if (match) return match;
  }
  return null;
}

/** Read the optional asset for a server-rendered profile, never throwing. */
export async function readMapStatsEntry(
  manifest: PackManifest,
  kind: MapStatsKind,
  name: string,
): Promise<MapStatsLookup | null> {
  const payload = await readPackJson<unknown>(manifest, PLAYER_MAP_STATS_PATH)
    .then(parsePlayerMapStats)
    .catch(() => null);
  if (!payload) return null;
  const entry = findMapStatsEntry(payload[kind], name, kind);
  if (!entry) return null;
  return { entry, window_days: payload.window_days, map_limit: payload.map_limit };
}

/** Read the optional asset for an /api/chat route. Throws only on transport. */
export async function lookupMapStats(
  kind: MapStatsKind,
  name: string,
): Promise<MapStatsLookup | null> {
  // Deliberately does NOT forward the route's deadline signal: the
  // signal-bearing branch of readPackJson bypasses the per-release cache, so
  // every request would download, hash and parse the whole multi-megabyte
  // artifact to answer for one subject. The cached branch serves repeats from
  // memory keyed by release; the route's overall deadline still bounds the
  // response through secureChatRoute's race.
  const payload = parsePlayerMapStats(
    await readChatJson<unknown>(PLAYER_MAP_STATS_PATH),
  );
  if (!payload) return null;
  const entry = findMapStatsEntry(payload[kind], name, kind);
  if (!entry) return null;
  return { entry, window_days: payload.window_days, map_limit: payload.map_limit };
}

File: lol_kills/export/player_map_stats.py
"""Per-map contribution statistics for the public player and team profiles.

The rows here are the *same* census the published ratings are fit on: this
module consumes the canonical player archive frame that
:func:`lol_kills.export.pack_records.build_profile_records` reads, so a profile
can never show a map the ratings did not see.

FAIL CLOSED.  Every metric is emitted as ``null`` when its source column is
absent or unusable.  A missing metric is never coerced to ``0``, because a
zero-value CS/min or damage share reads as a real (terrible) performance rather
than as an absent measurement.

Field naming is load bearing.  ``supabase/migrations/
20260815060001_descriptive_draft_query_api.sql`` installs
``public.scryglass_json_has_draft_fields``, which rejects any published asset
carrying a banned key.  ``gold`` is a banned *exact* key and ``strength_``,
``phase_``, ``live_``, ``control_`` and ``r9e_`` are banned prefixes, so the
descriptive names below (``gold_per_min``, ``gold_share_pct``, ...) are the only
safe spellings.  ``tests/test_player_map_stats.py`` reimplements the SQL ban
list and asserts every emitted key against it.

For the same reason ``players`` and ``teams`` are *arrays carrying a ``name``
field* rather than the name-keyed objects the neighbouring artifacts use.  The
guard inspects JSON object **keys**, so a name-keyed map would hand player and
team names to it: a roster with a player called ``Gold`` or an org tagged ``EV``
would silently fail publication with no signal pointing back at the roster. No
current identity collides, but nothing stops one appearing, and a name is data
we do not control.  As values, names are never inspected.
"""

from __future__ import annotations

from typing import Any, Mapping

import pandas as pd

from lol_kills.export.pack_records import (
    PUBLIC_ROLE_ALIASES,
    public_team_affiliation,
)
from lol_kills.etl.source_keys import canonical_source_game_key


SCHEMA_VERSION = "scryglass:player-map-stats:v1"

# Matches ``build_profile_records(recent_window_days=120)`` so the Stats tab and
# the rest of the profile describe the same window of play.
DEFAULT_WINDOW_DAYS = 120

# Per-identity row cap.  Measured against the real census, the 120-day window
# holds a median of 17 maps per player, so a cap of 20 keeps the typical player
# whole while trimming the stacked academy/main-roster tail. The whole artifact
# is loaded and parsed by the profile pages, and an uncapped build measures
# 13.1MB against a ~15MB budget; at 20 it measures 9.5MB.
#
# The aggregate header is computed over exactly the rows that survive this cap,
# so the header always describes the table printed beneath it.
DEFAULT_MAP_LIMIT = 20

_PLAYER_METRIC_KEYS = (
    "cs_per_min",
    "gold_per_min",
    "gold_share_pct",
    "damage_per_min",
    "damage_share_pct",
    "kda",
)

_TEAM_METRIC_KEYS = (
    "gold_per_min",
    "damage_per_min",
    "kills",
    "deaths",
    "game_length_min",
)


def _round(value: Any, digits: int = 2) -> float | None:
    """Round to a JSON-safe float, or fail closed to ``None``."""

    if value is None:
        return None
    try:
        if pd.isna(value):
            return None
    except (TypeError, ValueError):
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    if number != number or number in (float("inf"), float("-inf")):
        return None
    return round(number, digits)


def _positive(value: float | None) -> float | None:
    """Keep only strictly positive denominators."""

    return value if value is not None and value > 0 else None


def _mean(values: list[float | None]) -> float | None:
    """Mean over the *available* observations only, or ``None``."""

    present = [value for value in values if value is not None]
    if not present:
        return None
    return round(sum(present) / len(present), 2)


def _empty_payload(window_days: int, map_limit: int) -> dict[str, Any]:
    return {
        "schema_version": SCHEMA_VERSION,
        "window_days": window_days,
        "map_limit": map_limit,
        "players": [],
        "teams": [],
    }


def _prepared_frame(players: pd.DataFrame, window_days: int) -> pd.DataFrame:
    """Restrict the archive to complete, windowed, identifiable player maps."""

    frame = players.copy()
    identity_source = frame.get("game_uid", frame.get("gameid"))
    fallback = frame["gameid"] if "gameid" in frame.columns else None
    frame["_game_id"] = [
        canonical_source_game_key(value, fallback.loc[index] if fallback is not None else None)
        for index, value in identity_source.items()
    ]
    frame["_date"] = pd.to_datetime(frame["date"], utc=True, errors="coerce")
    frame["_side"] = frame["side"].astype(str).str.title()
    frame["_role"] = frame["position"].astype(str).str.casefold().map(PUBLIC_ROLE_ALIASES)
    frame["_player"] = frame["playername"].astype("string").str.strip()
    frame["_team"] = frame["teamname"].astype("string").str.strip()
    frame["_win"] = pd.to_numeric(frame["result"], errors="coerce")
    frame = frame[
        frame["_game_id"].astype(str).str.strip().ne("")
        & frame["_date"].notna()
        & frame["_side"].isin({"Blue", "Red"})
        & frame["_role"].notna()
        & frame["_player"].notna()
        & frame["_player"].ne("")
        & frame["_team"].notna()
        & frame["_team"].ne("")
        & frame["_win"].isin({0, 1})
    ].copy()
    if frame.empty:
        return frame
    latest = frame["_date"].max()
    if pd.notna(latest):
        frame = frame[frame["_date"].ge(latest - pd.Timedelta(days=window_days))].copy()
    return frame


def _side_gold_totals(frame: pd.DataFrame) -> dict[tuple[str, str], float | None]:
    """Team gold per (game, side), or ``None`` when the side is not complete.

    A gold share computed against a partial side is wrong rather than merely
    imprecise, so an incomplete side yields an unavailable share.
    """

    if "totalgold" not in frame.columns:
        return {}
    totals: dict[tuple[str, str], float | None] = {}
    gold = pd.to_numeric(frame["totalgold"], errors="coerce")
    for key, index in frame.groupby(["_game_id", "_side"], sort=False).groups.items():
        values = gold.loc[index]
        if len(values) != 5 or values.isna().any():
            totals[(str(key[0]), str(key[1]))] = None
            continue
        total = float(values.sum())
        totals[(str(key[0]), str(key[1]))] = total if total > 0 else None
    return totals


def _opponents(frame: pd.DataFrame) -> dict[tuple[str, str], str | None]:
    """Map (game, side) to the other side's display team name."""

    by_side: dict[str, dict[str, str]] = {}
    for (game_id, side), group in frame.groupby(["_game_id", "_side"], sort=False):
        names = {str(value) for value in group["_team"].dropna().unique()}
        if len(names) != 1:
            continue
        by_side.setdefault(str(game_id), {})[str(side)] = names.pop()
    opponents: dict[tuple[str, str], str | None] = {}
    for game_id, sides in by_side.items():
        for side in ("Blue", "Red"):
            other = "Red" if side == "Blue" else "Blue"
            opponents[(game_id, side)] = public_team_affiliation(sides.get(other))
    return opponents


def build_player_map_stats(
    players: pd.DataFrame,
    *,
    window_days: int = DEFAULT_WINDOW_DAYS,
    map_limit: int = DEFAULT_MAP_LIMIT,
) -> dict[str, Any]:
    """Build the per-map Stats payload for every windowed player and team."""

    if window_days < 1 or map_limit < 1:
        raise ValueError("window_days and map_limit must be positive")
    required = {"playername", "teamname", "side", "position", "result", "date"}
    if players is None or players.empty or not required.issubset(players.columns):
        return _empty_payload(window_days, map_limit)
    if "gameid" not in players.columns and "game_uid" not in players.columns:
        return _empty_payload(window_days, map_limit)

    frame = _prepared_frame(players, window_days)
    if frame.empty:
        return _empty_payload(window_days, map_limit)

    side_gold = _side_gold_totals(frame)
    opponents = _opponents(frame)

    numeric = {
        column: pd.to_numeric(frame[column], errors="coerce")
        if column in frame.columns
        else pd.Series(index=frame.index, dtype="float64")
        for column in (
            "cspm", "dpm", "damageshare", "totalgold", "gamelength",
            "kills", "deaths", "assists",
        )
    }
    has_champion = "champion" in frame.columns
    has_league = "league" in frame.columns

    player_rows: dict[str, list[dict[str, Any]]] = {}
    team_rows: dict[str, dict[str, dict[str, Any]]] = {}

    for index, row in frame.iterrows():
        game_id = str(row["_game_id"])
        side = str(row["_side"])
        minutes = _positive(_round(numeric["gamelength"].loc[index] / 60.0, 4))
        total_gold = _round(numeric["totalgold"].loc[index], 4)
        side_total = side_gold.get((game_id, side))
        kills = _round(numeric["kills"].loc[index], 0)
        deaths = _round(numeric["deaths"].loc[index], 0)
        assists = _round(numeric["assists"].loc[index], 0)
        damage_share = _round(numeric["damageshare"].loc[index], 6)
        won = int(row["_win"]) == 1

        entry = {
            "game_id": game_id,
            "date": row["_date"].date().isoformat(),
            "league": str(row["league"]) if has_league and pd.notna(row["league"]) else None,
            "opponent": opponents.get((game_id, side)),
            "champion": str(row["champion"]) if has_champion and pd.notna(row["champion"]) else None,
            "position": str(row["_role"]),
            "win": won,
            "cs_per_min": _round(numeric["cspm"].loc[index]),
            "gold_per_min": (
                _round(total_gold / minutes) if total_gold is not None and minutes else None
            ),
            "gold_share_pct": (
                _round(total_gold / side_total * 100.0)
                if total_gold is not None and side_total
                else None
            ),
            "damage_per_min": _round(numeric["dpm"].loc[index]),
            "damage_share_pct": _round(damage_share * 100.0) if damage_share is not None else None,
            "kda": (
                _round((kills + assists) / max(deaths, 1.0))
                if kills is not None and assists is not None and deaths is not None
                else None
            ),
            "game_length_min": _round(minutes) if minutes else None,
        }
        player_rows.setdefault(str(row["_player"]), []).append(entry)

        team = public_team_affiliation(row["_team"])
        if team is None:
            continue
        bucket = team_rows.setdefault(team, {})
        aggregate = bucket.get(game_id)
        if aggregate is None:
            aggregate = {
                "game_id": game_id,
                "date": entry["date"],
                "league": entry["league"],
                "opponent": entry["opponent"],
                "win": won,
                "game_length_min": entry["game_length_min"],
                "_gold": [],
                "_damage": [],
                "_kills": [],
                "_deaths": [],
            }
            bucket[game_id] = aggregate
        aggregate["_gold"].append(total_gold)
        aggregate["_damage"].append(
            _round(numeric["dpm"].loc[index] * minutes, 4)
            if minutes and _round(numeric["dpm"].loc[index]) is not None
            else None
        )
        aggregate["_kills"].append(kills)
        aggregate["_deaths"].append(deaths)

    players_payload = [
        _identity_payload(name, rows, _PLAYER_METRIC_KEYS, map_limit)
        for name, rows in sorted(player_rows.items())
    ]
    teams_payload = [
        _identity_payload(
            team,
            [_finalize_team_row(row) for row in bucket.values()],
            _TEAM_METRIC_KEYS,
            map_limit,
        )
        for team, bucket in sorted(team_rows.items())
    ]
    return {
        "schema_version": SCHEMA_VERSION,
        "window_days": window_days,
        "map_limit": map_limit,
        "players": players_payload,
        "teams": teams_payload,
    }


def _finalize_team_row(row: dict[str, Any]) -> dict[str, Any]:
    """Collapse a team's five player rows into one team-level map row."""

    gold = [value for value in row.pop("_gold") if value is not None]
    damage = [value for value in row.pop("_damage") if value is not None]
    kills = [value for value in row.pop("_kills") if value is not None]
    deaths = [value for value in row.pop("_deaths") if value is not None]
    minutes = row.get("game_length_min")
    # A partial side would understate a team total, so only a complete side of
    # five measured players yields a team sum.
    row["gold_per_min"] = (
        _round(sum(gold) / minutes) if len(gold) == 5 and minutes else None
    )
    row["damage_per_min"] = (
        _round(sum(damage) / minutes) if len(damage) == 5 and minutes else None
    )
    row["kills"] = int(sum(kills)) if len(kills) == 5 else None
    row["deaths"] = int(sum(deaths)) if len(deaths) == 5 else None
    return row


def _identity_payload(
    name: str,
    rows: list[dict[str, Any]],
    metric_keys: tuple[str, ...],
    map_limit: int,
) -> dict[str, Any]:
    """Aggregate header plus the most recent capped map rows."""

    ordered = sorted(
        rows,
        key=lambda row: (str(row.get("date") or ""), str(row.get("game_id") or "")),
        reverse=True,
    )[:map_limit]
    maps = len(ordered)
    wins = sum(1 for row in ordered if row.get("win") is True)
    payload: dict[str, Any] = {
        "name": name,
        "maps": maps,
        "wins": wins,
        "losses": maps - wins,
        "win_rate": round(wins / maps, 4) if maps else None,
    }
    for key in metric_keys:
        payload[key] = _mean([row.get(key) for row in ordered])
    payload["games"] = ordered
    return payload


def player_map_stats_row_count(payload: Mapping[str, Any]) -> int:
    """Total emitted map rows across both identity kinds."""

    total = 0
    for kind in ("players", "teams"):
        for entry in payload.get(kind) or []:
            total += len(entry.get("games") or [])
    return total
